extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/DriverHolder.java

```diff
@@ -0,0 +1,690 @@
+      }
+
+      if (!nextPartitionOffsets.keySet().equals(startOffsets.keySet())) {
+        throw new ISE(
+            "WTF?! Restored partitions[%s] but expected partitions[%s]",
+            nextPartitionOffsets.keySet(),
+            startOffsets.keySet()
+        );
+      }
+    }
+    committerSupplier = new Supplier<Committer>()
+    {
+      @Override
+      public Committer get()
+      {
+        final Map<Integer, Long> snapshot = ImmutableMap.copyOf(nextPartitionOffsets);
+
+        return new Committer()
+        {
+          @Override
+          public Object getMetadata()
+          {
+            return ImmutableMap.of(
+                KafkaIndexTask.METADATA_NEXT_PARTITIONS, new KafkaPartitions(
+                    topic,
+                    snapshot
+                )
+            );
+          }
+
+          @Override
+          public void run()
+          {
+            // Do nothing
+          }
+        };
+      }
+    };
+
+    // remove partition for which offsets have been consumed fully
+    for (Map.Entry<Integer, Long> partitionOffset : nextPartitionOffsets.entrySet()) {
+      if (endOffsets.get(partitionOffset.getKey()).equals(partitionOffset.getValue())) {
+        logInfo("[Start-Job] read partition [%d]", partitionOffset.getKey());
+        assignment.remove(partitionOffset.getKey());
+      }
+    }
+    if (assignment.isEmpty()) {
+      // finish the driver
+      logInfo("[Start-Job] read all partitions");
+      Preconditions.checkState(
+          driverStatus == DriverStatus.OPEN,
+          "WTH?! Cannot change driver status to [%s] from [%s]",
+          DriverStatus.COMPLETE,
+          driverStatus
+      );
+      driverStatus = DriverStatus.COMPLETE;
+    }
+
+    return nextPartitionOffsets;
+  }
+
+  boolean canHandle(ConsumerRecord<byte[], byte[]> record)
+  {
+    return driverStatus == DriverStatus.OPEN
+           && endOffsets.get(record.partition()) != null
+           && record.offset() >= startOffsets.get(record.partition())
+           && record.offset() < endOffsets.get(record.partition());
+  }
+
+  public SegmentIdentifier add(InputRow row) throws IOException
+  {
+    Preconditions.checkState(driverStatus == DriverStatus.OPEN, "Cannot add to driver which is not open!");
+
+    SegmentIdentifier identifier = driver.add(row, sequenceName, committerSupplier, false);
+
+    if (identifier == null) {
+      // Failure to allocate segment puts determinism at risk, bail out to be safe.
+      // May want configurable behavior here at some point.
+      // If we allow continuing, then consider blacklisting the interval for a while to avoid constant checks.
+      throw new ISE("Could not allocate segment for row with timestamp[%s]", row.getTimestamp());
+    }
+
+    // remember to call incrementNextOffsets before using any other method
+    // which relies on updated value of nextPartitionOffsets
+    if (driver.numRowsInSegment(identifier) >= driver.getMaxRowPerSegment()) {
```

Comment:
```driver.getMaxRowPerSegment()``` looks that each driver has different maxRowsPerSegment which is actually not. It would be better to use ```KafkaTuningConfig.getMaxRowsPerSegment()``` instead and remove ```driver.getMaxRowPerSegment()```.